File: 2-2/전공/알고리즘/과제/코딩과제1-정렬알고리즘/shellsort/c/shellsort.c
#include <stdio.h>
#define MAX_SIZE 1000
void ShellSort(int a[], int n);
int main()
{
    int numTestCases;
    scanf_s("%d", &numTestCases);
    for (int i = 0; i < numTestCases; ++i)
    {
        int num;
        int a[MAX_SIZE];
        scanf_s("%d", &num);
        for (int j = 0; j < num; j++)
            scanf_s("%d", &a[j]);
        ShellSort(a, num);
    }
    return 0;
}
void swap(int* a, int* b)
{
    int tmp = *a;
    *a = *b;
    *b = tmp;
}
/* Shell Sort 함수 */
void ShellSort(int A[], int n)
{
    int countCmpOps = 0; // 비교 연산자 실행 횟수
    int countSwaps = 0; // swap 함수 실행 횟수
    // Shell sort 알고리즘 구현
    int shrinkRatio = 2;
    int gap = n / shrinkRatio;
    while (gap > 0)
    {
        for (int i = gap; i < n; i++)
        {
            int tmp = A[i];
            int j = i;
            for (j = i; (j >= gap); j -= gap)
            {
                countCmpOps++;
                if (A[j - gap] > tmp)
                {
                    countSwaps++;
                    A[j] = A[j - gap];
                }
                else { break; }
            }
            A[j] = tmp;
        }
        gap /= shrinkRatio;
    }
    printf("%d %d ", countCmpOps, countSwaps);
}

File: 2-2/전공/알고리즘/과제/코딩과제1-정렬알고리즘/shellsort/shellsort.py
import sys
def ShellSort(A , n):
    shrinkRatio = 2
    gap = int(n / shrinkRatio)
    compare_cnt=0
    swap_cnt=0
    while( gap > 0 ):
        i = gap
        while(i < n):
            tmp = A[i]
            j = i
            while(True):
                if(j < gap):
                    break
                compare_cnt+=1
                if(A[j - gap] > tmp):
                    swap_cnt+=1
                    A [j] = A[j - gap]
                    j -= gap
                else :
                    break
            A [j] = tmp
            i += 1

        gap /= shrinkRatio
        gap = int(gap)
    print(compare_cnt, " ", swap_cnt)

n = int(sys.stdin.readline())

for i in range(n):
    arr = list(map(int, sys.stdin.readline().split()))
    m = arr[0]
    arr = arr[1:]
    ShellSort(arr, m)
